upb/mini_table.c

```diff
@@ -884,34 +962,71 @@ upb_MiniTable* upb_MiniTable_BuildMapEntry(upb_FieldType key_type,
   return ret;
 }
 
-upb_MiniTable_Extension* upb_MiniTable_BuildExtensions(const char* data,
-                                                       size_t len,
-                                                       size_t* ext_count,
-                                                       upb_Arena* arena,
-                                                       upb_Status* status) {
+upb_MiniTable_Enum* upb_MiniTable_BuildEnum(const char* data, size_t len,
+                                            upb_Arena* arena,
+                                            upb_Status* status) {
   upb_MtDecoder decoder = {
-      .arena = arena,
       .status = status,
-      .table = NULL,
   };
 
-  upb_MiniTable_Extension* exts;
+  if (UPB_SETJMP(decoder.err)) {
+    return NULL;
+  }
+
+  upb_MiniTable_Enum* table = upb_Arena_Malloc(arena, sizeof(*table));
+  upb_MtDecoder_CheckOutOfMemory(&decoder, table);
+
+  table->mask = 0;
+  table->value_count = 0;
+  table->values = NULL;
+
+  const char* ptr = data;
+  const char* end = UPB_PTRADD(data, len);
+
+  // Currently we do minimal validation of invariants (eg. that values are in
+  // order).  We may want to add these, but more likely, we will want the format
+  // to be a more compact variation where these errors are not possible.
+  while (end - ptr >= 4) {
+    uint32_t val;
+    memcpy(&val, ptr, 4);
+    if (val >= 64) break;
+    table->mask |= 1ULL << val;
+    ptr += 4;
+  }
+
+  if (ptr != end) {
+    size_t bytes = end - ptr;
+    if (bytes / 4 * 4 != bytes) {
+      upb_MtDecoder_ErrorFormat(&decoder, "Bytes should be a multiple of 4");
+      UPB_UNREACHABLE();
+    }
+    table->values = upb_Arena_Malloc(arena, end - ptr);
+    upb_MtDecoder_CheckOutOfMemory(&decoder, table);
+    memcpy((void*)table->values, ptr, end - ptr);
```

Comment:
you shouldn't need this `(void*)` cast I think

Reply:
It's casting away const.